proof/invariant-abstract/ARM/ArchLevityCatch_AI.thy

```diff
@@ -56,16 +56,45 @@ lemma asid_high_bits_of_add:
   done
 
 lemma preemption_point_success [simp,intro]:
-  "((Inr (), s') \<in> fst (preemption_point s)) \<Longrightarrow>
-  \<exists>f es. s' = s \<lparr> machine_state := machine_state s \<lparr> irq_state := f (irq_state (machine_state s)) \<rparr>, exst := es \<rparr>"
-  apply (auto simp: in_monad preemption_point_def do_machine_op_def
-                    select_f_def select_def getActiveIRQ_def alternative_def
-                    do_extended_op_def OR_choiceE_def mk_ef_def
-             split: option.splits if_splits
-             intro: exI[where x=id])
-      apply (rule_tac x=Suc in exI, rule_tac x="exst bb" in exI, force)+
-    apply (rule_tac x=id in exI, rule_tac x="exst b" in exI, force)+
-    done
+  "\<lbrakk>((Inr (), s') \<in> fst (preemption_point s))\<rbrakk>
+   \<Longrightarrow> \<exists>f g es lasttime curtime consumed.
+              s' = s\<lparr>machine_state :=
+                     machine_state s\<lparr>irq_state := f (irq_state (machine_state s)),
+                                     time_state := g (time_state (machine_state s)),
+                                     last_machine_time := lasttime\<rparr>,
+                     cur_time := curtime,
+                     consumed_time := consumed,
+                     exst := es \<rparr>"
+  apply (clarsimp simp: in_monad preemption_point_def do_machine_op_def
+                        select_f_def select_def getActiveIRQ_def alternative_def
+                        do_extended_op_def OR_choiceE_def mk_ef_def
+                        get_sc_refill_sufficient_def get_sched_context_def return_def
+                        refill_sufficient_def get_object_def get_sc_active_def
+                        update_time_stamp_def getCurrentTime_def andM_def ifM_def
+                 split: if_split_asm)
+  apply safe
+        apply ((case_tac x; clarsimp simp: fail_def return_def;
+               rule_tac x=Suc in exI, rule_tac x=Suc in exI, rule_tac x="exst bb" in exI,
+               rule_tac x="of_nat (min (18446744073709551615 - (unat kernelWCET_ticks + 3 * unat MAX_PERIOD))
```

Comment:
Done! I've go an input abbreviation in `MachineOps.thy`, and I've moved the abbreviation for `max_time` there too

Reply:
So because the input abbreviation is in an arch specific file, it can be seen in the Arch locale, but not in the generic files. And it doesn't seem like any fact is produced, so I can't add anything to `requalify_` in `MachineExports.thy`.

I thought it should be in the arch specific files, because we wanted the `type_synonym`s for `ticks` and `time` to be arch specific (we discussed this a month or so ago). 

Any thoughts?

Reply:
Ignore me: this can be added to `requalify_consts`